proposals/js-string-builtins/Overview.md

```diff
@@ -12,66 +12,65 @@ However, the overhead of importing glue code is prohibitive for primitives such
 
 These two pieces in combination allow runtimes to reliably emit optimal code sequences for JavaScript String operations within WebAssembly modules. In the future, other builtin objects or primitives can be exposed through new builtins.
 
-## Compile-time imports
+## Do we need builtins?
 
-Today, imports are provided when instantiating a module. This prevents web runtimes from being able to know anything about an import (beyond the type specified in the module) while compiling, without either speculation or deferring compilation until instantiation or later. This makes it difficult to optimize for a specific import when compiling a module.
+It is already possible today to import JS builtin functions (such as String.prototoype.getCharCodeAt) from wasm modules. Instead of defining new wasm specific-builtins, we could just re-use those directly.
 
-Speculation and deferred compilation are useful techniques, but should not be necessary to efficiently use specific JavaScript primitives from WebAssembly code on the Web.
+There are several problems with this approach.
 
-This proposal modifies the WebAssembly JS-API methods for compilation to optionally accept an object that specifies certain import values earlier than instantiation. There are several constraints to keep in mind before fully describing the API.
+The first problem is that existing API’s require a calling convention conversion to handle differences around the `this` value, which WebAssembly function import calls leave as `undefined`. The second problem is that certain primitive use JS operators such as `===` and `<` that cannot be imported. A third problem is that most JS builtins are extremely permissive of the types of values they accept, and it's desirable to leverage wasm's type system to remove those checks and coercions wherever we can.
 
-### Constraints
+It seems that creating new importable definitions that adapt existing JS primitives to WebAssembly is simpler and more flexible in the future.
 
-#### Modules can be shared across web workers using postMessage
+## Do we need builtin modules?
 
-Not all import values are shareable across workers. We must be able to send shareable import values across workers, and reject unshareable values.
+There is a variety of execution techniques for WebAssembly. Some WebAssembly engines compile modules eagerly (at WebAssembly.compile), some use interpreters and dynamic tiering, and some use on-demand compilation (after instantiation) and dynamic tiering.
 
-It’s possible that we could disable sharing modules that have any compile-time imports for an initial version, but this would need to be solved in the fullness of time.
+If we just have builtin functions, it would be possible to normally import then without any work to add builtin modules. The main issue is that imported values are not known until instantiation, and so engines that compile eagerly would be unable to generate specialized code to these imports.
```

Comment:
```suggestion
If we just have builtin functions, it would be possible to normally import them without any work to add builtin modules. The main issue is that imported values are not known until instantiation, and so engines that compile eagerly would be unable to generate specialized code to these imports.
```